<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Комментарий ревьюера</b>
    
Привет! Меня зовут Светлана Носова, я буду проверять твой проект. Если не против, давай общаться на «ты». Если тебе такой стиль общения не нравится, напиши об этом, будем на «вы». 

Сразу хочу обратить внимание на цветовую разметку.
Зеленым отмечены отличные решения и идеи, которые можно и нужно использовать в будущем. Желтым отмечено то, что можно или нужно оптимизировать, в зависимости от контекста. 
    
**Ответы на мои комментарии или доработки по моим рекомендациям попрошу выделить каким-нибудь цветом, кроме перечисленных.**
Также я попрошу тебя **не менять, не передвигать и не удалять мои комментарии**, чтобы при последующей проверке мне было проще ориентироваться.
    
Кроме того, комментарии у меня обозначены как заголовки. Это сбивает содержание, но при последующей проверке я могу снять с них заголовок, если тебе так будет удобнее. При этом они удобны, поскольку ты сразу можешь перейти к ним.   
        
<hr>
    
**По проекту:** работа выполнена очень хорошо, все оформлено понятно и аккуратно, есть анализ и промежуточные выводы. У меня остались некоторые вопросы, я все написала в комментариях. Там же я оставила рекомендации по улучшению проекта. Буду ждать работу на повторное ревью.
    
<hr>
Комментарии можешь оставлять с помощью такого кода в Markdown ячейке:
    
    
    <div class="alert alert-info">
    <h2> Комментарий студента</h2>

    Тут текст.
    </div>
  
    
Если не хочешь, чтобы твои комментарии отображались в содержании как заголовки, замени **h2** на **b**.    
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Добрый день! Не знаю, видно ли вам или нет, меня зовут Александр Кондрашов. Я не против общаться на "ты" и даже за, если вы будете обращаться ко мне на "ты", но мне привычнее к вам обращаться на "вы". Потом может быть привыкну и буду называть вас на "ты" :)
    
Немного не понимаю, как тут менять цвет ячеек, в Jupiter notebook...

Заголовки меня не напрягают.
    
Спасибо! Далее постарался ответить на вопросы и исправить содержание работы.
Единственное, что я не совсем понял, как быть с исправлениями в ячейке кода: я решил просто переписать код там, где оно требуется, "закомментировав" мой предыдущий код. Так нормально?
Исправления в выводах вы сможете заметить по заголовкам в формате: "Вывод (исправлено)"

</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Комментарий ревьюера 2</b>
    
Я оставила новые комментарии с цифрой 2, а со старых сняла заголовки, чтобы  не мешали. 
    
Код можно оставлять закоментированным, а можно и удалять. В первом случае старайся не создавать слишком много комментариев. Или выноси в отдельную ячейку. Все-таки ты учишься и это твои решения, так что полезно оставить. Потом уже можешь удалить. Для ревьюеров можешь писать, мол, для себя эту кучу комментариев оставил. 
    

Вот несколько слов об оформлении ячеек:    
</div>

<font color='red'> Так ты можешь выделять текст цветом. </font> 



<div style="border: 3px solid green; padding: 10px">
<b> Комментарий мученика науки </b>
    
Кликли дважды на ячейку, чтобы увидеть код. 
    
- border: 3px solid green; Это параметры границы (border), ее толщина в пикселях и цвет.  
    
- padding: 10px; Расстояние от рамки до ее содержимого, в пикселях.    
</div>

<div style="border-radius: 15px; background-color: mistyrose; padding: 15px;">
<b> Комментарий</b>
    
- border-radius: 15px; Добавили радиус, что округлило края ячейки. Чем выше число, тем сильнее округляем. 
    
- background-color: salmon; Добавили цвет. Не забывай точку с запятой после каждого параметра.
    
    
Цвета можешь найти [здесь](https://matplotlib.org/stable/gallery/color/named_colors.html). 
</div>

<div style="border-radius: 15px; background-color: mistyrose; border: 3px solid indigo; padding: 15px;">
<b> Комментарий</b>
    
Заливку и цвет границы можно можно делать разными.
    
    
Оставлю ссылки на статьи, где ты можешь поближе познакомиться с возможностями Markdown ячейки: [раз](https://sqlbak.com/blog/jupyter-notebook-markdown-cheatsheet) [два](https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd).    
    
Надеюсь, мои комментарии были полезны :) Работу я могу принять, поскольку вопросов больше нет. Поздравляю с первым решенным проектом и желаю успехов в дальнейшем обучении :)
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

**Изучение общей информации**

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.info()
data.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Всего - 21525 значений.
В столбцах days_employed и total_income есть пропуски в значениях, тип данный вещественный, а не целочисленный.

**Образование**

In [2]:
data['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Есть 5 категорий.
Данные необходимо перевести в нижний регистр, дабы они были пригодны для анализа.
Информация об уровне образования клиентов для последующего анализа и ответа на вопросы нам НЕ нужна.

**Семейное положение**

In [3]:
data.groupby('family_status')['family_status_id'].value_counts()

family_status          family_status_id
Не женат / не замужем  4                    2813
в разводе              3                    1195
вдовец / вдова         2                     960
гражданский брак       1                    4177
женат / замужем        0                   12380
Name: family_status_id, dtype: int64

Есть 5 категорий.
Определенному семейному статусу соответствует свой family_status_id.
С данными все в понядке, обрабатывать не нужно.
Данные понадобятся нам для ответа на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?".

**Тип занятости**

In [4]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

Есть 8 категорий.
Информация о типе занятости клиентов для последующего анализа и ответа на вопросы нам НЕ нужна.

**Ежемесячный доход**

Данные о ежемесячном доходе необходимо перевести в числовой формат и классифицировать по группам.
Данные понадобятся нам для ответа на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?".

**Наличие и количество детей**

In [5]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Данные понадобятся нам для ответа на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?".
Данные необходимо разделить на соответствующие группы (есть дети/нет) дабы сделать из них выводы.
Возьмем, что все значения, отличные от нуля означают наличие у клиента ребенка.

**Цель получения кредита**

In [6]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Вышеизложенное множество категорий в столбце purpose необходимо лемматизировать и сузить пул категорий.
Данные понадобятся нам для ответа на вопрос "Как разные цели кредита влияют на его возврат в срок?".

**Наличие задолженности**

In [7]:
data['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

Данные содержут информацию о наличии задолженности: 0 - задолженности не было, 1 - задолженность была.
Данные прямым образом нам понадобятся для ответа на вопросы кейса. 

**Вывод**

Вышеизложенные данные содержат всю необходимую информацию для ответа на вопросы кейса. 
Часть из них необходимо обработать и привести в вид, пригодный для анализа.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Отличная вводная часть.
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Спасибо :)
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Пропуски данных есть только в столбцах days_employed (19351 non-null float64 из 21525) и total_income (19351 non-null float64 из 21525).
Столбец days_employed нам не нужен, так как не является репрезентативным для нашего исследования.
Целесообразно заняться обработкой пропусков в столбце total_income, так как именно относительно него мы ответим на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?".

Для того, чтобы наше исследование было наиболее точным, необходимо заменить средним значением ЗП там, где она пропущена, связав ее с наименованием типа занятости.

In [8]:
data[data['total_income'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Везде, где total_income = NaN, days_employed тоже NaN.

In [9]:
print(len(data[data['total_income'].isna()]))

print(len(data[data['total_income'].isna()])/ 21525 * 100)

2174
10.099883855981417


Всего пропусков 2174, это 10% от общего числа значений.

Доля достаточно велика, так что стоит обработать пропуски.
Найдем среднее значение ЗП по типам занятости.

In [10]:
#income_count = data.groupby('income_type')['total_income'].count()
#income_sum = data.groupby('income_type')['total_income'].sum()
income_avg = data.groupby(['income_type', 'gender', 'family_status']).agg({'total_income': ['count', 'sum', 'mean']})
income_avg

total_income                \
                                                    count           sum   
income_type     gender family_status                                      
безработный     F      гражданский брак                 1  2.027225e+05   
                M      женат / замужем                  1  5.995699e+04   
в декрете       F      женат / замужем                  1  5.382913e+04   
госслужащий     F      Не женат / не замужем           85  1.425962e+07   
                       в разводе                       70  1.208359e+07   
                       вдовец / вдова                  40  6.241514e+06   
                       гражданский брак               190  2.898729e+07   
                       женат / замужем                577  8.819244e+07   
                M      Не женат / не замужем           55  1.054161e+07   
                       в разводе                        6  1.389952e+06   
                       вдовец / вдова                   1  2.682846e+05   
                       гражданский брак                52  1.082157e+07   
                       женат / замужем                236  5.143271e+07   
компаньон       F      Не женат / не замужем          442  8.132737e+07   
                       в разводе                      185  3.702205e+07   
                       вдовец / вдова                  80  1.440683e+07   
                       гражданский брак               588  1.069860e+08   
                       женат / замужем               1575  2.889743e+08   
                M      Не женат / не замужем          283  6.128988e+07   
                       в разводе                       58  1.453355e+07   
                       вдовец / вдова                   6  1.040582e+06   
                       гражданский брак               314  7.291997e+07   
                       женат / замужем               1045  2.477602e+08   
                XNA    гражданский брак                 1  2.039052e+05   
пенсионер       F      Не женат / не замужем          276  3.835930e+07   
                       в разводе                      184  2.488302e+07   
                       вдовец / вдова                 463  6.158425e+07   
                       гражданский брак               465  6.259302e+07   
                       женат / замужем               1418  1.886926e+08   
                M      Не женат / не замужем           37  5.437358e+06   
                       в разводе                       14  2.474665e+06   
                       вдовец / вдова                  21  2.435682e+06   
                       гражданский брак               111  1.767949e+07   
                       женат / замужем                454  6.799051e+07   
предприниматель F      гражданский брак                 1  4.991631e+05   
                M      женат / замужем                  0  0.000000e+00   
сотрудник       F      Не женат / не замужем          737  1.148207e+08   
                       в разводе                      404  6.382709e+07   
                       вдовец / вдова                 232  3.497553e+07   
                       гражданский брак              1314  1.982749e+08   
                       женат / замужем               3424  4.978045e+08   
                M      Не женат / не замужем          609  9.891921e+07   
                       в разводе                      162  2.782403e+07   
                       вдовец / вдова                  22  3.305703e+06   
                       гражданский брак               698  1.239800e+08   
                       женат / замужем               2412  4.523302e+08   
студент         M      Не женат / не замужем            1  9.820163e+04   

                                                             
                                                       mean  
income_type     gender family_status                         
безработный     F      гражданский брак       202722.511368  
                M      женат / замужем        

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера 2</h2>
    
Можно даже из таблицы брать значения:
</div>

In [20]:
# КОД РЕВЬЮЕРА
income_avg.loc['сотрудник', 'F', 'в разводе']

total_income  count    4.040000e+02
              sum      6.382709e+07
              mean     1.579878e+05
Name: (сотрудник, F, в разводе), dtype: float64

In [21]:
# КОД РЕВЬЮЕРА
income_avg.loc['сотрудник', 'F', 'в разводе'][2]

157987.84321312813

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера 2</h2>
    
А дальше создать функцию, которая будет возвращать нужные элементы по заданным характеристикам. Но мне такой подход не нравится, потому что можно сделать гораздо изящнее: так, как сделал ты. Только можно одной строкой:
    
    
    data['total_income'] = data.groupby(['income_type','gender', 'family_status'])['total_income'].apply(lambda x: x.fillna(x.median()))
    
    
Оно же, но без лямбды:
    
    
    data['total_income'] = data['total_income'].fillna(df.groupby(['income_type','gender', 'family_status'])['total_income'].transform('median'))    
</div>

Сгруппируем значения для удобства замены NaN средними значениями среднемесячного дохода, исходя из типов занятости.

In [11]:
grouped = data.groupby(['income_type', 'gender', 'family_status'])
data['total_income'] = grouped['total_income'].apply(lambda x: x.fillna(x.mean()))
#data.info()
data['total_income'].fillna(data.groupby(['income_type'])['total_income'].mean())

#print(data[data['income_type'] == 'предприниматель']['total_income'].mean())
data['total_income'].fillna(data[data['income_type'] == 'предприниматель']['total_income'].mean(), inplace = True)

print(data['total_income'].isna().sum())
#print(data[data['total_income'].isna() == True])

0


<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Отлично. Раз уж у нас есть такая возможность, давай попробуем заменить медианой (или средним), но взятой не по одной, а по 2-3 характеристикам клиента?
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Сгруппировал данные по соответствующим характеристикам клиента: тип занятости, пол и семейный статус. Почему? 
Целей кредита очень много, а данные об образовании еще предстоит обработать, это самые "стабильные" столбцы.
    
В процессе группировки возник казус: группа из одного человека, среднее значение которой получается NaN.
Принял решение заменить этот NaN на среднее значение среди группы income_type = предприниматель.
</div>

**Общий трудовой стаж в днях**

Сам по себе столбец для анализа нам не нужен, но NaN, видимо, нужно как-то обработать.

In [12]:
data[data['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

     total_income                                purpose  
12  159274.685915                        сыграть свадьбу  
26  217935.194673                            образование  
29  138982.974546       строительство жилой недвижимости  
41  152846.514280       сделка с подержанным автомобилем  
55  134608.645346                        сыграть свадьбу  
65  216572.006163  операции с коммерческой недвижимостью  
67  133069.514006                покупка жилья для семьи  
72  217935.194673  операции с коммерческой недвижимостью  
82  145386.828679                                  жилье  
83  187533.265136                                  жилье

Для дальнейшей обработки данных и изменения их типа достаточно заменить NaN на нули, так как, помимо NaNов, к самим данным есть вопросы, - усреднять и обрабатывать их нет смысла.

In [13]:
data['days_employed'] = data['days_employed'].fillna(0)
#data.info()
data['days_employed'].isna().sum()

0

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Хорошо. А что вообще можешь сказать про этот столбец? Почему могли возникнуть пропуски, откуда отрицательные и огромные значения?</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

См. исправленный вывод.
</div>

**Вывод** (исправленный)

Обработанные данные ныне не содержат пропусков.

Сами пропуски могли возникнуть из-за того, что у человека нет официального рабочего стажа: был трудоустроен неофициально без записи в трудовую книжку или же человек не работал, или данные были не указаны, так как их нет.
Отрицательные значения могли возникнуть из-за ошибок заполнения данной информации, например она могла вводиться в формате "количество детей - 1" и данное "-" это как минус.
Огромные значения могли появиться из-за неверного формата ввода: значения были введены не в днях, а в часах/минутах/секундах.

Можно приступать к следующим этапам предобработки данных.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера 2</h2>
    
Можно заметить, что отрицательные значения стажа по модулю выглядят нормально. Думаю, это не минус, а тире. Огромные значения это загадка. Могут быть часами или пропусками, которые при выгрузке автоматически преобразовались в выбивающиеся значения. Сами пропуски это скорее всего ошибка. При неофициальном трудоустройстве можно выписку из банка взять, а стаж приравнять к нулю. Не думаю, что кто-то пойдет брать кредит, не имея дохода. </div>

### Замена типа данных

Не понимаю зачем, но надо заменить тип данных в столбцах с вещественного на целочисленный. Для этого используем способ изменения типа данных при помощи функции astype().

In [14]:
data['total_income'] = data['total_income'].astype('int64')
data['days_employed'] = data['days_employed'].astype('int64')
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод**

Тип данных всех численных значений - целочисленный. Так удобнее для дальнейшего анализа и ответа на вопросы.

### Обработка дубликатов

In [15]:
data.duplicated().sum()

54

Если не найти дубликаты, то анализ данных может привести к некорректным результатам.

Исходя из предыдущих действий, можно заметить, что в графе образования есть дубликаты, свзяанные с разным регистром строковых данных.
Исправим это.

In [16]:
data['education'] = data['education'].str.lower()
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Посчитаем количество дубликатов и далее уже примем решение, надо ли от них избавляться или нет (исходя из того, насколько сильно их наличие может повлиять на результаты нашего исследования).

In [17]:
data.drop_duplicates()
data.reset_index(inplace = True)
data.duplicated().sum()

0

**Вывод**

Поиск дубликатов "рядовыми" и простыми способами выдал всего 71 дубликат. Дубликаты были удалены с помощью функции drop_duplicates(). Данные могли задвоиться именно из-за разного регистра вводимых данных (потому что до приведения к нижнему регистру дубликатов было 54, а стало 71), а также из-за прочих ошибок ввода данных или повторного ввода данных одних и тех же клиентов (например категория женщин, сменивших фамилию после свадьбы). Далее имеет смысл провести лемматизацию данных столбика purpose.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Поскольку у нас нет уникальных идентификаторов пользователей, эти дубликаты вполне могут быть разными людьми. А вот дубликаты записей в столбцах вроде **Среднее** образование и **среднее** образование удалить надо было (путем приведения к одному регистру), это верно. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Я правильно понял, что если комментарий отмечен зеленым цветом, то мне дополнять и исправилять ничего не надо?
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Комментарий ревьюера 2</h2>
    
Да :)</div>

### Лемматизация

Определим уникальные значения столбца purpose.

In [18]:
purpose_unique = pd.unique(data['purpose'])
print(purpose_unique)

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


Лемматизация уникальных значений столбца purpose.

In [19]:
from pymystem3 import Mystem
m = Mystem()

lemmas = []
for i in purpose_unique:
    lemma = m.lemmatize(i)
    lemmas.append(lemma)

lemmas_united = [j for i in lemmas for j in i]

from collections import Counter
Counter(lemmas_united)

Counter({'покупка': 10,
         ' ': 59,
         'жилье': 7,
         '\n': 38,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 1,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'подержанный': 1,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1})

**Вывод**

Исходя из лемматизированных данных мы можем четко определить категории целей получения кредита.

### Категоризация данных

Принцип категоризации данных: Сужение спектра категорий в столбце purpose за счет объединения похожих, исходя из лемматизации и стемминга.

In [20]:
def purpose_change(purpose):
    lemmas_row = m.lemmatize(purpose)
    for i in lemmas_row:
        if 'авто' in i:
            return 'автомобиль'
        if 'недвиж' in i:
            return  'недвижимость'
        if 'свад' in i:
            return  'свадьба'
        if 'образов' in i:
            return 'образование'
        if 'жиль' in i:
            return 'недвижимость'

data['purpose_grouped']= data['purpose'].apply(lambda x: purpose_change(x))     
data['purpose_grouped'].value_counts()
data.head(10)

index  children  days_employed  dob_years education  education_id  \
0      0         1          -8437         42    высшее             0   
1      1         1          -4024         36   среднее             1   
2      2         0          -5623         33   среднее             1   
3      3         3          -4124         32   среднее             1   
4      4         0         340266         53   среднее             1   
5      5         0           -926         27    высшее             0   
6      6         0          -2879         43    высшее             0   
7      7         0           -152         50   среднее             1   
8      8         2          -6929         35    высшее             0   
9      9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_grouped  
0               покупка жилья    недвижимость  
1     приобретение автомобиля      автомобиль  
2               покупка жилья    недвижимость  
3  дополнительное образование     образование  
4             сыграть свадьбу         свадьба  
5               покупка жилья    недвижимость  
6           операции с жильем    недвижимость  
7                 образование     образование  
8       на проведение свадьбы         свадьба  
9     покупка жилья для семьи    недвижимость

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Согласна. </div>

**Вывод**

Мы сузили спектр категорий целей получения кредита до 4ех в новом столбце purpose_grouped.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [21]:
data.head(10)

index  children  days_employed  dob_years education  education_id  \
0      0         1          -8437         42    высшее             0   
1      1         1          -4024         36   среднее             1   
2      2         0          -5623         33   среднее             1   
3      3         3          -4124         32   среднее             1   
4      4         0         340266         53   среднее             1   
5      5         0           -926         27    высшее             0   
6      6         0          -2879         43    высшее             0   
7      7         0           -152         50   среднее             1   
8      8         2          -6929         35    высшее             0   
9      9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_grouped  
0               покупка жилья    недвижимость  
1     приобретение автомобиля      автомобиль  
2               покупка жилья    недвижимость  
3  дополнительное образование     образование  
4             сыграть свадьбу         свадьба  
5               покупка жилья    недвижимость  
6           операции с жильем    недвижимость  
7                 образование     образование  
8       на проведение свадьбы         свадьба  
9     покупка жилья для семьи    недвижимость

In [22]:
def children_exist(children):
    if children != 0:
        return 'есть дети'
    else:
        return 'нет детей'


data['children_exist'] = data['children'].apply(lambda x: children_exist(x))

Выше можно видеть количество людей каждой категории. Далее, чтобы увидеть влияет ли наличие детей на возврат кредита в срок надо рассчитать долю должников в каждой категории. 

In [23]:
#data.groupby('children_exist')['debt'].count()
#data.groupby('children_exist')['debt'].sum()
#print('Процент должников среди категории клиентов без детей:', data.groupby('children_exist')['debt'].sum()[1] / data.groupby('children_exist')['debt'].count()[1] * 100)
#print('Процент должников среди категории клиентов с детьми:', data.groupby('children_exist')['debt'].sum()[0] / data.groupby('children_exist')['debt'].count()[0] * 100)

data.groupby('children_exist').agg({'debt':['count', 'sum', lambda x: str(round(x.mean()*100,2)) +'%' ]})

debt                 
                count   sum <lambda_0>
children_exist                        
есть дети        7376   678      9.19%
нет детей       14149  1063      7.51%

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Комментарий ревьюера</b>
    
Обрати внимание на то, что ты делишь сумму на количество. Это не что иное, как среднее. То есть это все можно было в одной группировке сделать:</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Очень понравился ваш код, сделаю по вашему примеру, спасибо за прекрасную идею!
</div>

In [24]:
# КОД РЕВЬЮЕРА
#data.groupby('children_exist').agg({'debt':['count', 'sum', lambda x: x.mean()*100]})

In [25]:
# КОД РЕВЬЮЕРА
#data.groupby('children_exist').agg({'debt':['count', 'sum', lambda x: str(round(x.mean()*100,2)) +'%' ]})

**Вывод**

Между наличием детей и возвратом кредита в срок есть зависимость: среди группы клиентов с детьми больше должников, чем среди группы клиентов без детей.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
def family_status_brak(family_status_id):
    if family_status_id == 0 or family_status_id == 1:
        return 'брак'
    else:
        return 'нет'


#data['family_status_brak'] = data['family_status_id'].apply(lambda x: family_status_brak(x))

#data.groupby('family_status_brak')['debt'].count()
#data.groupby('family_status_brak')['debt'].sum()

#print('Процент должников среди категории клиентов состоящих в браке:', data.groupby('family_status_brak')['debt'].sum()[0] / data.groupby('family_status_brak')['debt'].count()[0] * 100)
#print('Процент должников среди категории клиентов не состоящих в браке:', data.groupby('family_status_brak')['debt'].sum()[1] / data.groupby('family_status_brak')['debt'].count()[1] * 100)

data.groupby('family_status').agg({'debt':['count', 'sum', lambda x: str(round(x.mean()*100,2)) +'%' ]})

debt                
                       count  sum <lambda_0>
family_status                               
Не женат / не замужем   2813  274      9.74%
в разводе               1195   85      7.11%
вдовец / вдова           960   63      6.56%
гражданский брак        4177  388      9.29%
женат / замужем        12380  931      7.52%

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<b> Комментарий ревьюера </b>
    
Давай здесь рассмотрим все исходные категории. Например, развод может негативно влиять на погашение кредита из-за алиментов. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

Добавил и доработал.
</div>

**Вывод** (исправленный)

Между семейным положением и возвратом кредита в срок зависимость есть: доля должников среди холостяков и лиц в гражданском браки почти на 2 процента больше остальных групп.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<b> Комментарий ревьюера </b>
    
Даже если разница мала, все равно озвучивай все в выводе. Заказчик может неверно истолковать твои результаты, а потом обвинит аналитиков, если что-то у него пойдет не так. </div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

См. исправленный вывод
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
def income_category(total_income):
    if total_income <= 50000:
        return '1) 0-50k'
    if 50000 < total_income <= 100000:
        return '2) 50-100k'
    if 100000 < total_income <= 200000:
        return '3) 100-200k'
    if 200000 < total_income <= 300000:
        return '4) 200-300k'
    if total_income > 300000:
        return '5) 300k+'
    else:
        return 'иная категория'
    
data['income_category'] = data['total_income'].apply(lambda x: income_category(x))

#data.groupby('income_category')['debt'].count()
#data.groupby('income_category')['debt'].sum()

#print('Процент должников среди категории 0-50k:', data.groupby('income_category')['debt'].sum()[0] / data.groupby('income_category')['debt'].count()[0] * 100)
#print('Процент должников среди категории 50-100k:', data.groupby('income_category')['debt'].sum()[4] / data.groupby('income_category')['debt'].count()[4] * 100)
#print('Процент должников среди категории 100-200k:', data.groupby('income_category')['debt'].sum()[1] / data.groupby('income_category')['debt'].count()[1] * 100)
#print('Процент должников среди категории 200-300k:', data.groupby('income_category')['debt'].sum()[2] / data.groupby('income_category')['debt'].count()[2] * 100)
#print('Процент должников среди категории 300k+:', data.groupby('income_category')['debt'].sum()[3] / data.groupby('income_category')['debt'].count()[3] * 100)

data.groupby('income_category').agg({'debt':['count', 'sum', lambda x: str(round(x.mean()*100,2)) +'%' ]})

debt                 
                 count   sum <lambda_0>
income_category                        
1) 0-50k           372    23      6.18%
2) 50-100k        4091   331      8.09%
3) 100-200k      11764  1012       8.6%
4) 200-300k       3815   269      7.05%
5) 300k+          1483   106      7.15%

**Вывод**

Между уровнем дохода и возвратом кредита в срок есть зависимость: больше всего должников среди средней (по нашей классификации) категории.

- Как разные цели кредита влияют на его возврат в срок?

In [28]:
#data.groupby('purpose_grouped')['debt'].count()
#data.groupby('purpose_grouped')['debt'].sum()

#print('Процент должников среди категории автомобиль:', data.groupby('purpose_grouped')['debt'].sum()[0] / data.groupby('purpose_grouped')['debt'].count()[0] * 100)
#print('Процент должников среди категории недвижимость:', data.groupby('purpose_grouped')['debt'].sum()[1] / data.groupby('purpose_grouped')['debt'].count()[1] * 100)
#print('Процент должников среди категории образование:', data.groupby('purpose_grouped')['debt'].sum()[2] / data.groupby('purpose_grouped')['debt'].count()[2] * 100)
#print('Процент должников среди категории свадьба:', data.groupby('purpose_grouped')['debt'].sum()[3] / data.groupby('purpose_grouped')['debt'].count()[3] * 100)

data.groupby('purpose_grouped').agg({'debt':['count', 'sum', lambda x: str(round(x.mean()*100,2)) +'%' ]})

debt                
                 count  sum <lambda_0>
purpose_grouped                       
автомобиль        4315  403      9.34%
недвижимость     10840  782      7.21%
образование       4022  370       9.2%
свадьба           2348  186      7.92%

**Вывод** (исправленный)

Между целью кредита и возвратом кредита в срок есть зависимость: должников среди клиентов, цель которых взять вредит на автомобиль больше всего, далее по доле должников идет группы людей, берущих кредит на образование, проведение свадьбы, покупку недвижимости в порядке убывания.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<b> Комментарий ревьюера </b>
    
Тоже хороший пример. Я знак **>** воспринимаю как *больше* или как *лучше*. Так, не видя цифр (или не понимая их) я могу истолковать вывод так: по авто возвращают лучше, чем по кредитам за образование. За образование лучше, чем за свадьбу. В нашем коде никто разбираться не будет, будет просто презентация или отчет. Надо приучивать себя формулировать выводы и строить графики так, чтобы все было понятно без лишних вопросов. 
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

См. исправленный вывод
</div>

## Шаг 4. Общий вывод

С помощью предобработки данных у нас получилось привести данные в вид, пригодный для анализа.

1) Самые надежные клиенты это люди, не имеющие детей (7.51% процент невозврата в срок), а среди клиентов с детьми процент должников больше (9.19%).
2) Самые надежные клиенты это вдовцы и люди, находящиеся в разводе (6.56% и 7.11% процент невозврата в срок соответственно), а самые безответственные - холостяки и лица, живущие в гражданском браке(9.74% и 9.29% соответственно).
3) Самые надежные клиенты это люди с уровнем дохода до 50к рублей и 200к+ (6.18% и 7.10% процент невозврата в срок соответственно), а самые безответственные - люди с уровнем дохода 100-200к(8.6%). 
4) самые надежные клиенты это люди, берущие кредит на операции с недвижимостью (7.21% процент невозврата в срок), а самые безответственные те, кто обращается в банк с целью получить деньги на покупку автомобиля (9.34%).

Итого, идеальный портрет клиента. согласно нашим входным данным:
- не имеет детей
- вдовец/резведен
- зарабатывает 200к+
- собирается покупать недвижимость

Портрет безответственного лица:
- имеет детей
- холостяк
- зарабатывает 100-200к
- собирается покупать автомобиль


<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid">
<b> Комментарий ревьюера</b>
    
Хорошо. В идеале надо бы привести проценты невозврата в срок у самых надежных и самых безответственных клиентов для каждой группы. Например: самые надежные клиенты это люди, берущие кредит на операции с недвижимостью (такой-то процент невозврата в срок), а самые безответственные те, кто обращается в банк с целью получить деньги на покупку автомобиля (такой-то процент). И так для каждый группы. Это хороший тон, — подкреплять вывод полученными ранее результатами. Кроме того, так ты не оставляешь заказчику шансов неверно интерпретировать вывод.
</div>

<div class="alert alert-info">
<h2> Комментарий студента</h2>

См. исправленный вывод
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.